src/porepy/utils/tangential_normal_projection.py

```diff
@@ -193,74 +175,114 @@ def project_normal(self, num=None):
+                vector.
 
-    def _decompose_vector(self, nc):
-        if self.dim == 3:
-            t1 = np.random.rand(self.dim, 1) * np.ones(self.num_vecs)
-            t2 = np.random.rand(self.dim, 1) * np.ones(self.num_vecs)
-            normal, tc1, tc2 = self._gram_schmidt(nc, t1, t2)
-            basis = np.hstack([tc1, tc2, normal])
-        else:
-            t1 = np.random.rand(self.dim, 1) * np.ones(self.num_vecs)
-            normal, tc1 = self._gram_schmidt(nc, t1)
-            basis = np.hstack([tc1, normal])
-        return basis, normal
+            np.ndarray: ``shape=(dim, num_vecs)``. The normal vector.
 
-    def _gram_schmidt(self, u1, u2, u3=None):
         """
-        Perform a Gram Schmidt procedure for the vectors u1, u2 and u3 to obtain a set of
-        orthogonal vectors.
 
-        Parameters:
-            u1: ndArray
-            u2: ndArray
-            u3: ndArray
+        if self.dim == 2:
+            # Normalize the normal vector, just to be sure
+            normal = nc / np.linalg.norm(nc, axis=0)
+            # The tanegntial vector in 2d is deterministic up to an arbitrary sign. We
+            # choose the sign such that the vector points in the positive x-direction.
+            tc1 = np.zeros_like(normal)
+            negative_n1 = normal[1] < 0
+            tc1[:, negative_n1] = np.vstack(
+                [-normal[1, negative_n1], normal[0, negative_n1]]
+            )
+            positive_n1 = np.logical_not(negative_n1)
+            tc1[:, positive_n1] = np.vstack(
+                [normal[1, positive_n1], -normal[0, positive_n1]]
+            )
 
-        Returns:
-            u1': ndArray u1 / ||u1||
-            u2': ndarray (u2 - u2*u1 * u1) / ||u2||
-            u3': (optional) ndArray (u3 - u3*u2' - u3*u1')/||u3||
-        """
-        u1 = u1 / np.sqrt(np.sum(u1**2, axis=0))
-
-        u2 = u2 - np.sum(u2 * u1, axis=0) * u1
-        u2 = u2 / np.sqrt(np.sum(u2**2, axis=0))
+            basis = np.hstack([tc1, normal])
+        else:  # self.dim == 3
+            # Normalize the normal vector, just to be sure
+            normal = nc / np.linalg.norm(nc, axis=0)
+            # In 3d, there is some freedom in choosing the tangential vectors, but it is
+            # important that the definition minimize the risk for poorly conditioned
+            # computations (implementer note to self: Exactly what this means is
+            # unclear, but we have had spurious numerical issues that were traced back
+            # to a previous implementation using random vectors within the tangent
+            # plane. What caused this was never clear, but we do not want to take
+            # chances here). We choose one tangential vector to lie in the plane
+            # orthorgonal to the maximum direction of the normal vector (ex: a normal
+            # vector of [1, 3, 2] would give a first tangent vector in the xz-plane).
```

Comment:
To spell out what we mean by maximum component:
```suggestion
            # vector of [1, 3, 2] has maximum component in the y direction and thus gives a first tangent vector in the xz-plane).
```